File: python/ex3/oneVsAll.py
import scipy.optimize as opt
import numpy as np

from sigmoid import sigmoid
from lrCostFunction import lr_cost_function

def one_vs_all(X,Y,num_labels,lmd):
	# 给数据添加偏置维度
	X = np.c_[np.ones(X.shape[0]),X]
	n = X.shape[1]
	# 保存所有theta的集合
	all_theta = np.zeros((num_labels,n))
	# Y中的值是1~10
	for i in range(1,num_labels+1):
		init_theta = np.zeros((n,1));
		y = (Y == i).astype(int)
		
		def cost_func(t):
			return lr_cost_function(X,y,t,lmd)[0]
		def grad_func(t):
			return lr_cost_function(X,y,t,lmd)[1]

		theta, cost, *unused = opt.fmin_bfgs(f=cost_func, fprime=grad_func, x0=init_theta, maxiter=100, full_output=True, disp=False)
		all_theta[i-1,:] = theta.T

	return all_theta

File: matlab/ex3/oneVsAll.m
function [all_theta] = oneVsAll(X,Y,num_labels,lambda)
%% 函数功能：对多个类别的数据同时进行训练，迭代计算出每个类别的训练模型参数theta，返回所有theta的集合all_theta
[m,n] = size(X);

all_theta = zeros(num_labels , n+1);
X=[ones(m,1) , X];
options = optimset('GradObj', 'on', 'MaxIter', 50);

% 和前面所写的计算梯度的最大区别在于，这里是对多个theta向量进行计算，前面是对单个theta向量进行求解，所以需要加一个for循环
% 在计算labe-K的梯度代价和梯度时，是把其他类别都设置成0(Y==k)，转换成两分类来计算
for K =1:num_labels
    init_theta = zeros(n+1,1);
    costFun = @(t) lrCostFunction(X, (Y == K),t, lambda);
    % fmincg的用法和fminunc类似，但在处理大量参数时，更高效
    [theta] = fmincg(costFun , init_theta , options);  
    all_theta(K,:) = theta.';
end

end
